File: GOLANG/merge_sort.go
package main

import (
	"fmt"
	"math/rand"
 "time"
)

func MergeSort(arr []int, inicio int, fin int) {

   if ((fin - inicio) < 2) {
      return;
   }
   mid := (fin - inicio) / 2
   MergeSort(arr, inicio, (inicio + mid))
   MergeSort(arr, (inicio + mid), fin)
   Merge(arr, inicio, (inicio + mid), fin)
}

func Merge(arr []int, inicio int, mid int, fin int) {
   size, i, j := (fin - inicio), inicio, mid
   slice := make([]int, size, size)
   for k := 0; k < size; k++ {
      if i > (mid - 1) && j <= (fin - 1) {
         slice[k] = arr[j]
         j++
      } else if j > (fin - 1) && i <= (mid - 1) {
         slice[k] = arr[i]
         i++
      } else if arr[i] < arr[j] {
         slice[k] = arr[i]
         i++
      } else {
         slice[k] = arr[j]
         j++
      }
   }
   i = inicio;
   for k := 0; k < size; k++ {
      arr[i] = slice[k]
      i++
   }
}

func main() {
   //a := []int {4, 1, 5, 8, 9, 3, 2, 6, 7}
   var tam int
   rand.Seed(time.Now().UnixNano())
   pruebas := []int {100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000}
   //pruebas := []int {100}
   for j := 0; j < 15; j++ {
      tam = pruebas[j]
      fmt.Println("Para ", tam, ":")
      arr := make([]int, tam, tam)
      for i := 0; i < tam; i++ {
         arr[i] = rand.Intn(500000)
      }
      t0 := time.Now()
      MergeSort(arr, 0, tam)
      t1 := time.Now()
      fmt.Println(t1.Sub(t0))
   }
}

File: Python/merge_sort.py
import time
import datetime
from random import seed
from random import randint

def MergeSort(arr, inicio, fin):
    if ((fin - inicio) < 2):
       return
    mid = (fin - inicio) / 2
    MergeSort(arr, inicio, (inicio + mid))
    MergeSort(arr, (inicio + mid), fin)
    Merge(arr, inicio, (inicio + mid), fin)

def Merge(arr, inicio, mid, fin):
    size, i, j = (fin - inicio), inicio, mid
    sli = []
    for k in range (0, size):
        if i > (mid - 1) and j <= (fin - 1):
           sli.append(arr[j])
           j = j + 1
        elif j > (fin - 1) and i <= (mid - 1):
           sli.append(arr[i])
           i = i + 1
        elif arr[i] < arr[j]:
           sli.append(arr[i])
           i = i + 1
        else:
           sli.append(arr[j])
           j = j + 1
    i = inicio
    for k in range (0, size):
        arr[i] = sli[k]
        i = i + 1

def main():
    #a = [4, 1, 5, 8, 9, 3, 2, 6, 7]
    seed()
    pruebas = [100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000]
    #pruebas = [50000]
    for j in range (0, 15):
        tam = pruebas[j]
        print "Para ", tam, ":"
        arr = []
        for i in range (0, tam):
            arr.append(randint(0, 500000))
        #print arr
        t0 = datetime.datetime.now()
        inicio = time.time()
        MergeSort(arr, 0, tam)
        t1 = datetime.datetime.now()
        fin = time.time()
        procesamiento = fin - inicio
        td = t1 - t0
        print "Tiempo de ejecucion: " + format((procesamiento*1000), ".4f")
        print td.total_seconds()
        #print arr

main()
